rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/SuspendedState.java

```diff
@@ -0,0 +1,108 @@
+package org.rascalmpl.vscode.lsp.dap;
+
+import io.usethesource.vallang.ISourceLocation;
+import org.rascalmpl.debug.IRascalFrame;
+import org.rascalmpl.interpreter.Evaluator;
+import org.rascalmpl.interpreter.result.IRascalResult;
+import org.rascalmpl.uri.URIUtil;
+import org.rascalmpl.vscode.lsp.dap.variable.RascalVariable;
+import org.rascalmpl.vscode.lsp.dap.variable.VariableSubElementsCounter;
+import org.rascalmpl.vscode.lsp.dap.variable.VariableSubElementsCounterVisitor;
+import org.rascalmpl.vscode.lsp.dap.variable.VariableSubfieldsVisitor;
+
+import java.util.*;
+
+/**
+ * Class used to store the state of the Rascal Evaluator when it is suspended
+ */
+public class SuspendedState {
+
+    private final Evaluator evaluator;
+    private IRascalFrame[] currentStackFrames;
+    private final HashMap<Integer, RascalVariable> variables;
+    private final HashMap<Integer, IRascalFrame> scopes;
+    private int referenceIDCounter;
+
+    private boolean isSuspended;
+
+
+    public SuspendedState(Evaluator evaluator){
+        this.evaluator = evaluator;
+        this.variables = new HashMap<>();
+        this.scopes = new HashMap<>();
+    }
+
+    public void suspended(){
+        Stack<IRascalFrame> stack = evaluator.getCurrentStack();
+        currentStackFrames = stack.toArray(IRascalFrame[]::new);
+        referenceIDCounter = 0;
+        this.variables.clear();
+        this.scopes.clear();
+        this.isSuspended = true;
+    }
+
+    public ISourceLocation getCurrentLocation(){
+        return evaluator.getCurrentPointOfExecution() != null ?
+            evaluator.getCurrentPointOfExecution()
+            : URIUtil.rootLocation("stdin");
+    }
+
+    public void resumed(){
+        this.isSuspended = false;
+    }
+
+    public boolean isSuspended() {
+        return isSuspended;
+    }
+
+    public IRascalFrame[] getCurrentStackFrames(){
+        return currentStackFrames;
+    }
+
+    public IRascalFrame getCurrentStackFrame(){
+        return currentStackFrames[currentStackFrames.length - 1];
+    }
+
+    public int addScope(IRascalFrame frame){
+        scopes.put(++referenceIDCounter, frame);
+        return referenceIDCounter;
+    }
+
+    public List<RascalVariable> getVariables(int referenceID, int startIndex, int maxCount){
+        List<RascalVariable> variableList = new ArrayList<>();
+
+        if(referenceID < 0) return variableList;
```

Comment:
we prefer not to have ifs without braces, just to avoid us later adding a line and not realizing there were no curly braces.